explorer/frontend/scripts/regen.hs

```diff
@@ -0,0 +1,133 @@
+import System.IO (openFile, hClose, IOMode(WriteMode))
+import Filesystem.Path.CurrentOS (encodeString)
+
+main :: IO ()
+main = do
+  testMode <- options "Regenerate frontend dependencies" optionsParser
+  sh $ if testMode
+    then test
+    else regen
+
+regen :: Shell ()
+regen = do
+  echo "Regenerating nix for frontend"
+  bower2nix "bower.json" "bower-generated.nix"
+  node2nix "package.json" "node-packages.nix"
+
+bower2nix :: FilePath -> FilePath -> Shell ()
+bower2nix src out = cachedShell src out $ \out' ->
+  procs "bower2nix" [tt src, tt out'] empty
+
+node2nix :: FilePath -> FilePath -> Shell ()
+node2nix src out = cachedShell src out $ \out' -> do
+  let composition = "composition.nix"
+  out'' <- mktempfile "." (tt out)
+  composition' <- mktempfile "." (tt composition)
+  procs "node2nix" ["-6", "-i", tt src, "-c", tt composition', "-o", tt out, "-d"] empty
+  cp out out''
+  liftIO $ fixNix composition' composition
+  liftIO $ fixNix out'' out'
+  inplace uglify2 out'
+  inplace srcAttr out'
+  inplace addSrcArg composition
+
+test :: Shell ()
+test = do
+  echo "Checking that auto-generated frontend dependencies nix is up to date."
+  b <- needsChange "bower.json" "bower-generated.nix"
+  n <- needsChange "package.json" "node-packages.nix"
+  when b $ echo " - bower-generated.nix needs update"
+  when n $ echo " - node-packages.nix needs update"
+  when (b || n) $ die "Run explorer/frontend/scripts/regen.hs to fix this"
+
+optionsParser :: Parser Bool
+optionsParser = switch "test" 't' "Test freshness but don't regenerate"
+
+----------------------------------------------------------------------------
+
+-- | Run a shell command only if the destination file is out of date
+-- with respect to the source file.
+cachedShell :: FilePath -> FilePath -> (FilePath -> Shell ()) -> Shell ()
+cachedShell src dst action = needsChange src dst >>= \case
+  True -> do
+    printf ("Generating " % fp % "\n") dst
+    tmp <- mktempfile "." (tt $ basename dst)
+    action tmp
+    whenM (testpath dst) $ rm dst
+    input tmp & stampFile src & output dst
+  False -> printf (fp % " is already up to date according to " % fp % "\n") dst src
+
+-- | A file needs a change if its hash doesn't match or it doesn't exist.
+needsChange :: FilePath -> FilePath -> Shell Bool
+needsChange src dst = do
+  exists <- testfile dst
+  if exists
+    then do
+      line <- limit 1 $ input dst
+      hash <- limit 1 $ hashFile src
+      pure $ line /= hash
+    else pure True
+
+-- | sha256sum output prepended with a nix line comment symbol
+hashFile :: FilePath -> Shell Line
+hashFile src = fmap ("# " <>) (inproc "sha256sum" [tt src] empty)
+
+-- | Adds a hash to the top of the file
+stampFile :: FilePath -> Shell Line -> Shell Line
+stampFile ref f = cat [hashFile ref, f]
+
+----------------------------------------------------------------------------
+
+-- | Replace references to uglify version 3 with uglify version 2
+uglify2 :: Pattern Text
```

Comment:
Why can't this be handled with conversion itself? Maybe we just need a comment :) I fear this will break in a month :)

Reply:
Yes it's not very good. I couldn't find a way of forcing npm to always use a certain version (e.g. `yarn.lock` has version 3 in there also).

I have reworked to make the versions less hardcoded.

I have also generated a `package-lock.json`.

Hopefully these will reduce possibility of breakage.